src/codegen/expression.rs

```diff
@@ -1601,6 +1601,33 @@ fn expr_builtin(
 
             Expression::Undefined(tys[0].clone())
         }
+        ast::Builtin::WriteBytesData | ast::Builtin::WriteStringData => {
+            let buffer = expression(&args[0], cfg, contract_no, func, ns, vartab, opt);
+            let data = expression(&args[1], cfg, contract_no, func, ns, vartab, opt);
+            let offset = expression(&args[2], cfg, contract_no, func, ns, vartab, opt);
+
+            let advanced_ptr = Expression::AdvancePointer {
+                pointer: Box::new(buffer),
+                bytes_offset: Box::new(offset),
+            };
+
+            let size = Expression::Builtin(
+                Loc::Codegen,
+                vec![Type::Uint(32)],
+                Builtin::ArrayLength,
+                vec![data.clone()],
+            );
+
+            cfg.add(
+                vartab,
+                Instr::MemCopy {
+                    source: data,
+                    destination: advanced_ptr,
+                    bytes: size,
+                },
+            );
+            Expression::Undefined(tys[0].clone())
```

Comment:
Missing bounds check - is this writing to be beyond the end of the buffer. This also needs some tests.

Reply:
Tests are available in `tests/solana_tests/builtin.rs`

Reply:
Where are the test for bounds checks?